### ISSUE TITLE
Callbacks not firing reliably with uploadInBackground: true

### ISSUE THREAD

**Issue by wuservices** [filestack/filestack-js#58]
I opened a ticket about this with more info, but just in case this get to the right person faster or other people have this issue, I'm posting here too.

Basically when images upload in the background before the user clicks upload, there's no way to track what happens. Here's example of events after choosing 2 images and letting them upload all the way before the user clicks on upload.

As is, this makes it impossible for me to use `uploadInBackground` since I'll never know when the uploads are complete.

```
Events with uploadInBackground: true
====================================
onOpen
--- SELECT 2 FILES ---
onFileSelected with file 1
onFileSelected with file 2
--- WAIT while files are uploaded in the background, then click Upload ---
onUploadStarted
onClose


Events with uploadInBackground: false
====================================
onOpen
--- SELECT 2 FILES ---
onFileSelected with file 1
onFileSelected with file 2
--- CLICK Upload ---
onUploadStarted
onFileUploadStarted with file 1
onFileUploadStarted with file 2
onUploadProgress ...
onUploadProgress ...
onUploadProgress ...
onUploadProgress ...
onFileUploadFinished with file 1
onUploadComplete with file 1
onUploadProgress ...
onFileUploadFinished with file 2
onUploadComplete with file 2
onClose
```

**Comment by velveteer** [comment]
@wuservices You can know when uploads complete when `pick` resolves with the file metadata object, and this is the intended use case. Of course it won't be per file, only when all uploads have finished. Would it be preferable for callbacks to fire on background uploads by default? Right now you are correct in that they only fire for uploads started by the user. 

I don't remember why it was done this way, so unless I find a problem in that implementation I don't see any reason why these callbacks can't be respected for background uploads.

**Comment by wuservices** [comment]
@velveteer thanks for the info! Makes sense that you can use the Promise from `pick` to at the very least know when everything is done.

It would **definitely** be preferable to have all the callbacks fire for background uploads by default. It's very unexpected that `onFileUploadFinished` doesn't fire if it finished in the background. It's also not ideal that `onUploadProgress` doesn't fire for background uploads either, although at least that quirk was documented.

One thing that's missing vs the v2 uploader is a unique internal ID for each file. Adding this in (could be as simple as a counter), would allow you to correlate completion callbacks with files in progress. Currently, with v3, I make up my own ID with all the file info, but it's possible that if a user uploads 2 photos with the exact same info, I won't be able to tell them apart in the callbacks. One flaw with the v2 id system, was that the ids reset every time you re-opened the dialog, so you couldn't open up the uploader while uploads were still in progress and then continue to track additional files. The ID issue is easy to fix - just continue to increment the id for as long as the page is loaded instead of resetting it back when the dialog is reopened.

Requests:

- Make onUploadProgress fire for background uploads
- Make onFileUploadFinished fire for background uploads
- Add a unique ID for each file. It should be unique as long as the page is loaded.

**Here's my use case**
Overall, my goal is to make the smoothest, fastest, and most performant integration possible. Part of that includes fast upload performance (uploading ASAP in the background). It's also good to be able to show progress while photos are uploading (even while in the background), and to get notifications of when each file is done - incrementally, before every single file is done.

In my app, you can upload many images and use them to personalize something. Imagine a user picks 20 photos then clicks on Upload in the Filestack dialog. Ideally, they'd get a progress bar which grows as each file is uploaded. Each % of progress is based on the total number of bytes of that photo. For cloud sources, I approximate this based on dimensions, and fall back to an arbitrary size by default, since we don't always know the bytes. This gives semi-accurate progress, but it depends on onUploadProgress firing. For this many photos, we could hypothetically try to estimate time remaining as well, but that also depends on consistent progress updates.

Progress is a nice to have, but ultimately, the second each file is uploaded, we want to show the user each image in the app ASAP, so they can use it. However, we hit the metadata API and faces APIs in parallel before adding each photo to the app, to drive certain app logic.

As you could imagine, in this use case where somebody chooses many photos to use immediately, getting callbacks in the background is very important. As is, background uploading actually has somewhat of the opposite effect. For images that are uploaded in the background, before the user clicks on upload, we can't use them (or know about them) until the pick Promise is resolved, even though they're uploaded first.

**Comment by velveteer** [comment]
@wuservices What a coincidence because we added an `uploadId` field to file metadata for the next release (it will be a UUID attached to each file specifically for tracking them in callbacks). I will look into fixing the callback behavior with background uploads, and can probably have that in a fast follow.

Thanks for the details on your use case. Sounds like a perfect fit for our uploader, and we will work on the issues mentioned here to make it a better experience for you and your users.

**Comment by wuservices** [comment]
@velveteer that's amazing news! I love that you're so receptive to feedback and you seem to be thinking about all the right things already 👍 

Can't wait to launch an awesome integration. The only other thing we'd want is some access to the actual file, which now should be somewhat possible since the iframe is gone.

It'd be super cool and app-like to allow our users instantly use their photos using the file API and data URIs. You'd just have to give us access to the object, which you're already using for thumbnails anyways. I'm not sure how memory implications work in this scenario, but it seems like you've already thought of potential issues there since there was the flag to disable thumbnails.

**Comment by velveteer** [comment]
@wuservices 0.9.7 now includes a new option `exposeOriginalFile` which enables access to the actual File instance for local files. Callbacks should also now fire regardless of `uploadInBackground`. Let us know if this resolves your issue.

**Comment by wuservices** [comment]
@velveteer thanks for the quick iteration! I've been excited to try these changes out. The more consistent callbacks help out a lot by providing more regular uploads and they help remove a lot of special case code I had before to handle the previous inconsistencies. `globalDropZone` is a great addition! We had some users trying to drop images in and navigating away by accident. I like that the drop zone is the entire screen also.

One small issue. It seems like `uploadId` is `undefined` for `onFileUploadFailed`. This prevents us from resetting our global progress when there's an issue or when the user cancels a file.

`exposeOriginalFile` looks awesome! Any pointers around gotchas when working with the File API and big images? My first thought was to use the File reference to create small screen sized versions of each image to save memory vs using the original images, and then swap them out with the cloud versions once they're uploaded. Our goal would be to allow users to play use the images and interact with them while they're being uploaded in the background.

**Comment by velveteer** [comment]
@wuservices Glad to hear you're playing with the new stuff. 

I can't reproduce the missing `uploadId` on 0.9.7 with cloud or local files, but I am using a different config. If you're still seeing the issue on 0.9.7 can you provide your config so we can attempt to reproduce? 

Resizing image data with canvas is a popular way to go right now in browsers. The only gotcha may be performance, especially if you need to resize lots of images concurrently without locking up the main thread. If you're just dealing with a single large image you'll probably be fine using FileReader to grab its data, and optionally using canvas to resize it before displaying the preview (via object URL or data URL). There are some nice libraries, like pica, that have support for web workers to help reduce the lag of bulk resizing in the browser.

**Comment by wuservices** [comment]
@velveteer Thanks for the tips on resizing in the browser. Those are great pointers!

I see now that we do get `uploadId` for `onFileUploadFailed` when I kill the connection. However, perhaps the important detail is that I'm not getting it when a download is cancelled.

To reproduce, I pick a big local file, and I click the X by it before it finished uploading and before I click the Upload button. Then, `onFileUploadFailed` fires with `Error: Upload cancelled`, but `uploadId`, is `undefined`.

Besides `maxSize`, `maxFiles`, `accept`, `fromSources`, and callbacks, I'm using:
```
  hideWhenUploading: true
  uploadInBackground: true
  disableTransformer: true
  globalDropZone: true
```

Can you reproduce with that?

**Comment by wuservices** [comment]
@velveteer did you have any luck reproducing the issue with the missing `uploadId` with the steps above, by cancelling an upload in progress?

In case this helps somebody else, I'm going to work around the issue for now by preventing the user from canceling a background upload using CSS to hide the remove button:
```
.fsp-summary__action--remove {
  display: none;
}
```

**Comment by velveteer** [comment]
@wuservices Hey sorry for the delay, but so far I cannot reproduce the issue you are describing (using 0.9.8). I am logging the error and file within the `onFileUploadFailed` callback, and the `uploadId` shows up in the file object when I cancel it (in summary screen or by closing the modal).

**Comment by wuservices** [comment]
@velveteer Thanks for checking this out again. I upgraded to 0.9.8 and I'm not seeing this issue anymore. Awesome, all my wildest dreams have come true! Well, almost... Now if only your backend didn't take 10+ seconds to generate thumbnails after uploads complete, but that's a separate deal.

Really appreciate the quick responses to feedback here! Callbacks are good to go :)
